### PR TITLE
Support client results in sharding mode

### PR DESCRIPTION
**Pull request by xingsy97**
Support client result (also called "client invocation") in sharding mode (multiple endpoints)

Ref: Partial Support PR https://github.com/Azure/azure-signalr/pull/1828

### CODE REVIEW

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/MultiEndpointServiceConnectionContainer.cs [hunk 1]
```diff
@@ -231,13 +231,13 @@ internal IEnumerable<ServiceEndpoint> GetRoutedEndpoints(ServiceMessage message)
                     return _router.GetEndpointsForConnection(closeConnectionMessage.ConnectionId, endpoints);
 
                 case ClientInvocationMessage clientInvocationMessage:
-                    return SingleOrNotSupported(_router.GetEndpointsForConnection(clientInvocationMessage.ConnectionId, endpoints), clientInvocationMessage);
+                    return _router.GetEndpointsForBroadcast(endpoints);
```
Comment (JialinXin) [inline src/Microsoft.Azure.SignalR.Common/ServiceConnections/MultiEndpointServiceConnectionContainer.cs:234]: Shall we call `GetEndpointsForConnection`? Considering the client invocation should only send to the target client connection?

File: src/Microsoft.Azure.SignalR/ClientInvocation/CallerClientResultsManager.cs [hunk 2]
```diff
@@ -20,34 +23,54 @@ internal sealed class CallerClientResultsManager : ICallerClientResultsManager,
...
+        private IEndpointRouter? _endpointRouter { get; }
+        private IServiceEndpointManager? _serviceEndpointManager { get; }
+        private readonly AckHandler _ackHandler = new();
 
-        public CallerClientResultsManager(IHubProtocolResolver hubProtocolResolver)
+        public CallerClientResultsManager(IHubProtocolResolver hubProtocolResolver, IServiceEndpointManager? serviceEndpointManager = null, IEndpointRouter? endpointRouter = null)
         {
             _hubProtocolResolver = hubProtocolResolver ?? throw new ArgumentNullException(nameof(hubProtocolResolver));
+            if (serviceEndpointManager == null && endpointRouter != null) throw new ArgumentNullException(nameof(serviceEndpointManager));
+            if (serviceEndpointManager != null && endpointRouter == null) throw new ArgumentNullException(nameof(endpointRouter));
```
Comment (JialinXin) [inline src/Microsoft.Azure.SignalR/ClientInvocation/CallerClientResultsManager.cs:34]: serviceEndpointManager shouldn't be null.

File: src/Microsoft.Azure.SignalR/ClientInvocation/CallerClientResultsManager.cs [hunk 3]
```diff
@@ -20,34 +23,54 @@ internal sealed class CallerClientResultsManager : ICallerClientResultsManager,
...
+        public Task<T> AddInvocation<T>(string hub, string connectionId, string invocationId, CancellationToken cancellationToken)
         {
             var tcs = new TaskCompletionSourceWithCancellation<T>(
                 cancellationToken,
                 () => TryCompleteResult(connectionId, CompletionMessage.WithError(invocationId, "Canceled")));
 
+            var ackNumber = 1;
+            if (_serviceEndpointManager != null && _endpointRouter != null) {
+                var serviceEndpoints = _serviceEndpointManager.GetEndpoints(hub);
+                ackNumber = _endpointRouter.GetEndpointsForConnection(connectionId, serviceEndpoints).Count();
```
Comment (JialinXin) [inline src/Microsoft.Azure.SignalR/ClientInvocation/CallerClientResultsManager.cs:54]: This should be same as the one called in [MultiEndpointServiceConnectionContainer.cs](https://github.com/Azure/azure-signalr/pull/1852/files#diff-1f321e250ff8676cbfb17272b420c048bc70705b398a2e315fcbe28c76cb2f57)

File: src/Microsoft.Azure.SignalR/ClientInvocation/CallerClientResultsManager.cs [hunk 4]
```diff
@@ -101,21 +124,29 @@ public void CleanupInvocationsByConnection(string connectionId)
...
                 }
-
-                // if false the connection disconnected right after the above TryGetValue
-                // or someone else completed the invocation (likely a bad client)
-                // we'll ignore both cases
-                if (_pendingInvocations.TryRemove(message.InvocationId, out _))
+                
+                // Considering multiple endpoints, wait until 
+                // 1. Received a non-error CompletionMessage
+                // or 2. Receved messages from all endpoints
```
Comment (JialinXin) [inline src/Microsoft.Azure.SignalR/ClientInvocation/CallerClientResultsManager.cs:137]: typo: Rece*i*ved

File: src/Microsoft.Azure.SignalR/ClientInvocation/CallerClientResultsManager.cs [hunk 5]
```diff
@@ -20,28 +21,45 @@ internal sealed class CallerClientResultsManager : ICallerClientResultsManager,
...
+        private IEndpointRouter _endpointRouter { get; }
+        private IServiceEndpointManager _serviceEndpointManager { get; }
+        private readonly AckHandler _ackHandler = new();
 
-        public CallerClientResultsManager(IHubProtocolResolver hubProtocolResolver)
+        public CallerClientResultsManager(IHubProtocolResolver hubProtocolResolver, IServiceEndpointManager serviceEndpointManager, IEndpointRouter endpointRouter)
         {
             _hubProtocolResolver = hubProtocolResolver ?? throw new ArgumentNullException(nameof(hubProtocolResolver));
+            _serviceEndpointManager = serviceEndpointManager ?? throw new ArgumentNullException(nameof(serviceEndpointManager));
+            _endpointRouter = endpointRouter ?? throw new ArgumentNullException(nameof(endpointRouter));
```
Comment (JialinXin) [inline src/Microsoft.Azure.SignalR/ClientInvocation/CallerClientResultsManager.cs:32]: endpointRouter is nullable when it's single endpoint, as you rely on this to detect multi ack is needed in line 47...

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/MultiEndpointServiceConnectionContainer.cs [hunk 6]
```diff
@@ -231,13 +231,13 @@ internal IEnumerable<ServiceEndpoint> GetRoutedEndpoints(ServiceMessage message)
...
-                    return SingleOrNotSupported(_router.GetEndpointsForConnection(clientInvocationMessage.ConnectionId, endpoints), clientInvocationMessage);
+                    return _router.GetEndpointsForConnection(clientInvocationMessage.ConnectionId, endpoints);
 
                 case ServiceMappingMessage serviceMappingMessage:
-                    return SingleOrNotSupported(_router.GetEndpointsForConnection(serviceMappingMessage.ConnectionId, endpoints), serviceMappingMessage);
+                    return _router.GetEndpointsForBroadcast(endpoints);
 
                 case ServiceCompletionMessage serviceCompletionMessage:
-                    return SingleOrNotSupported(_router.GetEndpointsForConnection(serviceCompletionMessage.ConnectionId, endpoints), serviceCompletionMessage);
+                    return _router.GetEndpointsForBroadcast(endpoints);
```
Comment (JialinXin) [inline src/Microsoft.Azure.SignalR.Common/ServiceConnections/MultiEndpointServiceConnectionContainer.cs:240]: Should here send to connection as well?

File: src/Microsoft.Azure.SignalR/ClientInvocation/CallerClientResultsManager.cs (hunk 5 above)
Comment (xingsy97) [reply src/Microsoft.Azure.SignalR/ClientInvocation/CallerClientResultsManager.cs:32]: I should have removed detection in L47. `endpointRouter` is non-nullable.

**Review by JialinXin — APPROVED** [review]

**Pull request merged**